scheduler/src/cook/kubernetes/api.clj

```diff
@@ -0,0 +1,316 @@
+  (:require [clojure.tools.logging :as log]
+            [plumbing.core :as pc])
+  (:import
+    (com.google.gson JsonSyntaxException)
+    (com.twosigma.cook.kubernetes WatchHelper)
+    (io.kubernetes.client ApiClient ApiException)
+    (io.kubernetes.client.apis CoreV1Api)
+    (io.kubernetes.client.custom Quantity Quantity$Format)
+    (io.kubernetes.client.models V1Pod V1Container V1Node V1Pod V1Container V1ResourceRequirements V1EnvVar
+                                 V1ObjectMeta V1PodSpec V1PodStatus V1ContainerState V1DeleteOptionsBuilder
+                                 V1DeleteOptions)
+    (io.kubernetes.client.util Watch)
+    (java.util.concurrent Executors ExecutorService)))
+
+
+(def ^ExecutorService kubernetes-executor (Executors/newFixedThreadPool 2))
+
+(defn handle-watch-updates
+  [state-atom ^Watch watch key-fn callback]
+  (while (.hasNext watch)
+    (let [update (.next watch)
+          item (.-object update)
+          prev-item (get @state-atom (key-fn item))]
+      (case (.-type update)
+        "ADDED" (swap! state-atom (fn [m] (assoc m (key-fn item) item)))
+        "MODIFIED" (swap! state-atom (fn [m] (assoc m (key-fn item) item)))
+        "DELETED" (swap! state-atom (fn [m] (dissoc m (key-fn item)))))
+      (when callback
+        (try
+          (callback prev-item item)
+          (catch Exception e
+            (log/error e "Error while processing callback")))))))
+
+(defn initialize-pod-watch
+  "Initialize the pod watch. We require that this function set the current-pods-atom before it finishes so that
+  cook.kubernetes.compute-cluster/initialize-cluster works correctly."
+  [^ApiClient api-client current-pods-atom pod-callback]
+  (let [api (CoreV1Api. api-client)
+        current-pods (.listPodForAllNamespaces api
+                                               nil ; continue
+                                               nil ; fieldSelector
+                                               nil ; includeUninitialized
+                                               nil ; labelSelector
+                                               nil ; limit
+                                               nil ; pretty
+                                               nil ; resourceVersion
+                                               nil ; timeoutSeconds
+                                               nil ; watch
+                                               )
+        pod-name->pod (pc/map-from-vals (fn [^V1Pod pod]
+                                          (-> pod
+                                              .getMetadata
+                                              .getName))
+                                        (.getItems current-pods))]
+    (log/info "Updating current-pods-atom with pods" (keys pod-name->pod))
+    (log/error "TODO: We should have two atoms, one for cook pods (that feeds into the controller) and one for
+    all pods (that feeds into machine utilization then into offer generation)")
```

Comment:
Seems like this should be a comment rather than an ERROR log

Reply:
👍

Reply:
I want to make sure this isn't forgotten. Its a critical next change. We have 20? 30? TODO's that this could get lost in.

Reply:
Doesn't really matter to me. Let's just leave it.